### PR TITLE
feat: implement RFC 6750 Bearer token authentication

### PR DESCRIPTION
**Pull request by ThomasK33**
# Add RFC 6750 Bearer Token Authentication Support

This PR implements RFC 6750 Bearer Token authentication as an additional authentication method for Coder's API. This allows clients to authenticate using standard OAuth 2.0 Bearer tokens in two ways:

1. Using the `Authorization: Bearer <token>` header
2. Using the `access_token` query parameter

Key changes:

- Added support for extracting tokens from both Bearer headers and access_token query parameters
- Implemented proper WWW-Authenticate headers for 401/403 responses with appropriate error descriptions
- Added comprehensive test coverage for the new authentication methods
- Updated the OAuth2 protected resource metadata endpoint to advertise Bearer token support
- Enhanced the OAuth2 testing script to verify Bearer token functionality

These authentication methods are added as fallback options, maintaining backward compatibility with Coder's existing authentication mechanisms. The existing authentication methods (cookies, session token header, etc.) still take precedence.

This implementation follows the OAuth 2.0 Bearer Token specification (RFC 6750) and improves interoperability with standard OAuth 2.0 clients.

### CODE REVIEW

**Comment by ThomasK33** [comment]
* **#18747** <a href="https://app.graphite.dev/github/pr/coder/coder/18747?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18746** <a href="https://app.graphite.dev/github/pr/coder/coder/18746?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18712** <a href="https://app.graphite.dev/github/pr/coder/coder/18712?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18692** <a href="https://app.graphite.dev/github/pr/coder/coder/18692?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18670** <a href="https://app.graphite.dev/github/pr/coder/coder/18670?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18669** <a href="https://app.graphite.dev/github/pr/coder/coder/18669?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18646** <a href="https://app.graphite.dev/github/pr/coder/coder/18646?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18645** <a href="https://app.graphite.dev/github/pr/coder/coder/18645?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18644** <a href="https://app.graphite.dev/github/pr/coder/coder/18644?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/coder/coder/18644?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18643** <a href="https://app.graphite.dev/github/pr/coder/coder/18643?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18613** <a href="https://app.graphite.dev/github/pr/coder/coder/18613?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18575** <a href="https://app.graphite.dev/github/pr/coder/coder/18575?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18548** <a href="https://app.graphite.dev/github/pr/coder/coder/18548?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#18612** <a href="https://app.graphite.dev/github/pr/coder/coder/18612?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: coderd/httpmw/apikey.go [hunk 1]
```diff
@@ -214,6 +214,26 @@ func ExtractAPIKey(rw http.ResponseWriter, r *http.Request, cfg ExtractAPIKeyCon
...
+			switch {
+			case strings.Contains(response.Message, "invalid") || strings.Contains(response.Detail, "invalid"):
+				wwwAuth = `Bearer realm="coder", error="invalid_token", error_description="The access token is invalid"`
+			case strings.Contains(response.Message, "expired") || strings.Contains(response.Detail, "expired"):
+				wwwAuth = `Bearer realm="coder", error="invalid_token", error_description="The access token has expired"`
+			case strings.Contains(response.Message, "audience") || strings.Contains(response.Message, "mismatch"):
+				wwwAuth = `Bearer realm="coder", error="invalid_token", error_description="The access token audience does not match this resource"`
+			case code == http.StatusForbidden:
+				wwwAuth = `Bearer realm="coder", error="insufficient_scope", error_description="The request requires higher privileges than provided by the access token"`
+			}
```
Comment (johnstcn) [inline coderd/httpmw/apikey.go]: How about we map 401 to `invalid_token` and `403` to `insufficient_scope` instead? This is what the RFC recommends.

File: coderd/httpmw/rfc6750_extended_test.go
Comment (johnstcn) [inline coderd/httpmw/rfc6750_extended_test.go:1]: What's the significance of `extended` versus the regular test?

File: coderd/httpmw/rfc6750_extended_test.go [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+		wwwAuth := rec.Header().Get("WWW-Authenticate")
+		require.Contains(t, wwwAuth, "Bearer")
+		require.Contains(t, wwwAuth, "realm=\"coder\"")
+		require.Contains(t, wwwAuth, "error=\"invalid_token\"")
+		require.Contains(t, wwwAuth, "error_description=")
+	})
+}
+
+// TestOAuth2BearerTokenConcurrency tests Bearer token authentication under concurrent load
+func TestOAuth2BearerTokenConcurrency(t *testing.T) {
```
Comment (johnstcn) [inline coderd/httpmw/rfc6750_extended_test.go]: What is this actually testing? If we want to check for concurrency issues, we should be using `coder/quartz`. Also, the pattern of splitting up valid/invalid into subtests makes this whole thing dependent on `-parallel`.

File: coderd/httpmw/rfc6750_test.go [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+		// Check that WWW-Authenticate header is present
+		wwwAuth := rw.Header().Get("WWW-Authenticate")
+		require.NotEmpty(t, wwwAuth)
+		require.Contains(t, wwwAuth, "Bearer")
+		require.Contains(t, wwwAuth, `realm="coder"`)
+	})
+}
+
+// TestAPITokenFromRequest tests the RFC 6750 bearer token extraction directly
+func TestAPITokenFromRequest(t *testing.T) {
```
Comment (johnstcn) [inline coderd/httpmw/rfc6750_test.go:167]: suggestion: could be a table test

File: coderd/httpmw/rfc6750_test.go [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+		IPAddress:       defaultIPAddressForTests(),
+		LastUsed:        dbtime.Now(),
+		ExpiresAt:       dbtime.Now().Add(testutil.WaitLong),
+		CreatedAt:       dbtime.Now(),
+		UpdatedAt:       dbtime.Now(),
+		LoginType:       database.LoginTypePassword,
+		Scope:           database.APIKeyScopeAll,
+		LifetimeSeconds: int64(testutil.WaitLong.Seconds()),
+	})
+	require.NoError(t, err)
```
Comment (johnstcn) [inline coderd/httpmw/rfc6750_test.go]: Why not `dbgen.APIKey`?

File: coderd/httpmw/rfc6750_extended_test.go
Comment (ThomasK33) [reply coderd/httpmw/rfc6750_extended_test.go:1]: The idea was that  `rfc6750_test.go` contain basic RFC 6750 compliance tests (authentication methods, basic token validation), while the `rfc6750_extended_test.go` contain security and edge case testing (malformed headers, precedence, WWW-Authenticate compliance, etc.).

File: coderd/httpmw/rfc6750_extended_test.go (hunk 2 above)
Comment (ThomasK33) [reply coderd/httpmw/rfc6750_extended_test.go]: Yeah, you're right. I'll remove those. 👍

File: coderd/httpmw/apikey.go (hunk 1 above)
Comment (ThomasK33) [reply coderd/httpmw/apikey.go]: Yeah, I'll refactor this to switch on the status code first and then do the strings contains matching.

File: coderd/httpmw/rfc6750_test.go (hunk 4 above)
Comment (ThomasK33) [reply coderd/httpmw/rfc6750_test.go]: done

File: coderd/httpmw/rfc6750_test.go (hunk 3 above)
Comment (ThomasK33) [reply coderd/httpmw/rfc6750_test.go:167]: done

File: coderd/httpmw/rfc6750_test.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+// defaultIPAddressForTests returns a default IP address for test API keys
+func defaultIPAddressForTests() pqtype.Inet {
+	return pqtype.Inet{
+		IPNet: net.IPNet{
+			IP:   net.IPv4(127, 0, 0, 1),
+			Mask: net.IPv4Mask(255, 255, 255, 255),
+		},
+		Valid: true,
+	}
+}
```
Comment (johnstcn) [inline coderd/httpmw/rfc6750_test.go]: This shouldn't be needed anymore because you're using `dbgen`

```suggestion
```

**Review by johnstcn — APPROVED** [review]
Approving but would like a second pair of eyes.

File: coderd/httpmw/apikey.go [hunk 6]
```diff
@@ -214,6 +214,31 @@ func ExtractAPIKey(rw http.ResponseWriter, r *http.Request, cfg ExtractAPIKeyCon
...
+			var wwwAuth string
+
+			switch code {
+			case http.StatusUnauthorized:
+				// Map 401 to invalid_token with specific error descriptions
+				switch {
+				case strings.Contains(response.Message, "expired") || strings.Contains(response.Detail, "expired"):
+					wwwAuth = `Bearer realm="coder", error="invalid_token", error_description="The access token has expired"`
+				case strings.Contains(response.Message, "audience") || strings.Contains(response.Message, "mismatch"):
+					wwwAuth = `Bearer realm="coder", error="invalid_token", error_description="The access token audience does not match this resource"`
```
Comment (Emyrk) [inline coderd/httpmw/apikey.go:228]: Very unfortunate we have to do string matches :cry:

File: coderd/httpmw/apikey.go [hunk 7]
```diff
@@ -738,7 +768,18 @@ func APITokenFromRequest(r *http.Request) string {
...
+	authHeader := r.Header.Get("Authorization")
+	if strings.HasPrefix(authHeader, "Bearer ") {
+		return strings.TrimPrefix(authHeader, "Bearer ")
+	}
+
+	// Check access_token query parameter
+	accessToken := r.URL.Query().Get("access_token")
+	if accessToken != "" {
+		return accessToken
+	}
```
Comment (Emyrk) [inline coderd/httpmw/apikey.go]: You might want to add add the `Authorization` header auth to the CSRF exemption list? 

https://github.com/coder/coder/blob/119626d84d4c876b0522b73d94257de15209cb9a/coderd/httpmw/csrf.go#L26-L26

Basically cookie auth has to also send CSRF headers. Other header auths do not. This code just checks if the auth header is set and it matches the cookie, we still exempt the request. Super niche, but might avoid a customer report down the line.

**Review by Emyrk — APPROVED** [review]

**Comment by ThomasK33** [comment]
### Merge activity

* **Jul 2, 5:14 PM UTC**: @ThomasK33 merged this pull request with [Graphite](https://app.graphite.dev/github/pr/coder/coder/18644).
